rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,321 @@
+The initial implementation will only be supporting CSV files as a resource. It
+is anticipated that future work will expand the available resources to include
+other file types as well as databases.
+
+Like the services that sources and sinks integrate with, enrichment "tables"
+will likely need administrator approval and setup. Typically data is enriched
+from some sort of shared company "resource" that end users will likely not have
+access to. Admin will likely want to restrict access to resources to approved
+pipelines, especially if sensitive information is contained.
+
+The tables will need to be integrated with `vector validate`, which would
+ensure the resource exists and is correctly formatted.
+
+### Schema
+
+For the CSV Table all columns will be considered to be Strings. Since Tables are
+loaded before VRL compilation it will be possible to ensure that Vrl doesn't
+search on columns that do not exist within the datafile. Searching on a column
+that doesn't exist can prevent Vector from loading.
+
+If (when these features are implemented) the user attempts to reload a Table
+and a column that has been Indexed no longer exists, this should prevent the
+file from being loaded. Vector will continue to use the currently loaded data.
+
+These restrictions apply only to columns that VRL is searching on. If the user
+attempts to use a field that has not been returned from the enrichment, this
+will be possible, but the value will be Null. This is largely due to the way
+that VRL works with paths into Objects.
+
+### Vrl functions
+
+Two remap functions:
+
+#### `find_table_row`
+
+This function will look up a single row within the table dataset. If a single
+row is found that data is returned as an object, otherwise this function will
+error.
+
+A metric will be emitted to indicate the lookup time.
+
+#### `search_table_rows`
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the
+array.
+
+A metric will be emitted to indicate the lookup time.
+
+#### Parameters
+
+*table*
+
+The name of the table to lookup. This must point to a table specified in the
+config file eg `tables.csv_file`. Both functions are generic over all Table
+types.
+
+*condition*
+
+`condition` is a single level, key/value object that specifies the fields and
+values to lookup. The fields must all match (AND) for the row to be returned.
+
+*case_sensitive*
+
+By default the search will be case insensitive, but this can be changed by
+passing `true` to this parameter.
+
+### Example config
+
+```toml
+[tables.csv_file]
+    type = "csv"
+    file = "/path/to/csv.csv"
+    delimiter = ","
+
+[sources.datadog_logs]
+    type = "datadog_logs"
+    address = "0.0.0.0:80"
+
+[transforms.simple_enrich]
+    type = "remap"
+    inputs = ["datadog_logs"]
+    source = '''
+        . = parse_json!(.message)
+
+        result, err = find_table_row(
+            tables.csv_file,
```

Comment:
Do we expect to be able to look up things other than tables? If not, we could maybe shorted this to just `find_table_row(csv_file, ...)`.

Reply:
'tables' is just a standard VRL variable that will just be injected into the script. I think it would be better to namespace them all under a single variable `tables` to avoid conflicting with other variables the user may want to use.

Reply:
I'm not fond of magic variables (can people overwrite them, if not, we have to add some variable names to our list of reserved identifiers).

I do wonder how this intersects with event metadata and the desire to expose that information in VRL at some point.

I don't have a solution at hand, but either we reserve `tables` (and `metadata`) as reserved identifiers, or we find another way to allow operators to access special datasets that are separate from event fields and regular variables.

Reply:
People could overwrite `tables`, but if the script tried to use it in a call to `find_table_row` it would no longer compile since it would then be the wrong type. The type system should be sufficient to keep everything safe here.